### ISSUE TITLE
Enable freezing parts of the model in Fully Sharded Data Parallel

### ISSUE THREAD

**Issue by lxuechen** [pytorch/pytorch#83204]
### 🚀 The feature, motivation and pitch

Due to how `FlattenParamsWrapper` is used in the current FSDP implementation, there doesn't seem to be a straightforward way to shard a parameter which doesn't need to be optimized. 

The particular use case I'm aiming at is where one shards the whole model's parameters to save memory, while only computing gradients for a small subset of parameters. 

### Additional context

Related to https://github.com/pytorch/pytorch/issues/76501

@awgu

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501 @ezyang

**Comment by awgu** [comment]
Hi @lxuechen!

For the small subset of parameters that need gradients, are they in a small subset of the _modules_, or are they spread across many modules? If they are in a small subset of modules, with today's FSDP, you may be able to wrap those modules together in one `FullyShardedDataParallel` instance and set `requires_grad=False` for the FSDP's `FlatParameter`. If they are spread across many modules, then this will not work, and I am working on landing some changes this half (before end of year) to help with this.

The idea is that we actually expose the original parameters from `nn.Module.named_parameters()` (and hence `.parameters()`), so that the user passes the original parameters to the optimizer instead of the `FlatParameter`s. In that case, you can choose to set `requires_grad=False` for the original parameters you do not want to be optimized.

However, one caveat to mention is that from a memory usage perspective, the entire gradient for the `FlatParameter` is still materialized (even if only a small portion of the original parameters corresponding to the `FlatParameter` requires gradients), except FSDP will make sure to not forward any gradient to the original parameters that have `requires_grad=False`. In other words, this respects the desired behavior computationally but not memory-wise.

**Comment by lxuechen** [comment]
> For the small subset of parameters that need gradients, are they in a small subset of the modules, or are they spread across many modules?

There're unfortunately spread out in multiple modules. 

> However, one caveat to mention is that from a memory usage perspective, the entire gradient for the FlatParameter is still materialized (even if only a small portion of the original parameters corresponding to the FlatParameter requires gradients), except FSDP will make sure to not forward any gradient to the original parameters that have requires_grad=False. 

I can understand how this is extends the current implementation, but it's a little unfortunate one'd still need to materialize the entire grad due to using FlatParameter.

**Comment by lxuechen** [comment]
@awgu 

I just read through your [sketch](https://github.com/awgu/views/blob/753e4b9641407072e9e5af212ad23b2e90fa0cdf/views.py). 

It seems the lazy sol'n to the "freezing parts of the model" problem is really just to record the right offsets so one'd be able to index the right slices when performing updates (and zero'ing out the other slices which don't require grads, or just zero'ing out the grads as in your sketch). Is there something better we can do here? 

Is seems to me that to reach the desired memory behaviour, one'd need to have 2 separate `.flat_param` attributes recording and differentiating between the two types of parameters. This likely requires a re-design of the whole FSDP module.

**Comment by awgu** [comment]
The challenge with computing per-original parameter gradients comes from autograd.

The current implementation relies on the `FlatParameter` being the base `Tensor` involved in autograd, and each original parameter is a view into the `FlatParameter`, propagating its gradient into the `FlatParameter`'s gradient. This is useful in giving FSDP the right time to schedule the reduce-scatter since the `FlatParameter`'s gradient is only ready when all constituent parameters' gradients have been computed.

I can see two possibilities, one more likely than the other:
1. We support the case where one module can have more than one `FlatParameter`. In that case, you can wrap the parameters that do not require gradients into one `FlatParameter` and set its `requires_grad=False`, and you can wrap the remaining trainable parameters in a smaller `FlatParameter`. Hopefully, the less efficient communication of the smaller `FlatParameter` can still be amortized over the large `FlatParameter`. This is the more likely option.
2. The less likely option is to figure out a way such that we dissolve the `FlatParameter` and instead just enforce `k` original parameters to always be contiguous in memory such that they can be all-gathered together. Then, with respect to autograd, they are each individual `Tensor`s, where autograd does not see that the `k` parameters are contiguous. In other words, the contiguity is only useful for communication efficiency. In that way, the gradients are computed per original parameter. The issue here is how do we know when we can reduce-scatter their gradients, and we would need logic to determine whether we are reduce-scattering all of the gradients as one tensor or each gradient individually (which would be needed if we are freezing some parameters).

If your use case is really important, we can look into supporting option 1. I am working on refactoring FSDP to make this feasible down the line. Option 2 would require nontrivial (re)design.

Minor note: That sketch you linked is quite old, but I agree it does give the general idea.

**Comment by lxuechen** [comment]
I think having something like option 1 could be extremely helpful. Thanks for writing up the options here.

**Comment by lxuechen** [comment]
@awgu Speaking of the sketch, do you have a newer sketch where optimizers could take in regular parameters (not the flat ones)? 

I just want something quick'n dirty that I could iterate on for a moving project. 

Given that, sol'n 1 doesn't seem tremendously challenging to hack up, and I can also help with implementing it.

**Comment by lxuechen** [comment]
By the way, reading through your example again, I think I understand how it works. 

Here are a couple of questions: 
- Is it always necessary to create a `.grad` with all zeros [here](https://github.com/awgu/views/blob/753e4b9641407072e9e5af212ad23b2e90fa0cdf/views.py#L115)
- I haven't thought deeply, but your example should work with gradient accumulation, right? Do you see any potential gotchas? 
- Why does this `_acc_grad` variable need to be always recorded [here](https://github.com/awgu/views/blob/753e4b9641407072e9e5af212ad23b2e90fa0cdf/views.py#L141)?

**Comment by lxuechen** [comment]
Ok, since I didn't get a further response, I'll probably starting hacking the FSDP module from the latest stable release.

**Comment by rxqy** [comment]
Hi @awgu, is there any update on this issue?
A more common use case I can think of would be to set different lrs for different layers, e.g., no weight decay for norm layers. And this is a common practice in many [transformers architectures](https://huggingface.co/blog/pytorch-fsdp). I'm wondering if it is possible with the current flat params implementation?

**Comment by awgu** [comment]
> Hi @awgu, is there any update on this issue? A more common use case I can think of would be to set different lrs for different layers, e.g., no weight decay for norm layers. And this is a common practice in many [transformers architectures](https://huggingface.co/blog/pytorch-fsdp). I'm wondering if it is possible with the current flat params implementation?

I think different learning rates should be doable with today's nightlies and for the upcoming PT2.0 release. The way to do so is to pass `use_orig_params=True` to the FSDP constructor. Then, `named_parameters()` returns _sharded_ original parameters (i.e. whatever part of the parameter is on the rank), and you can pass different hyperparameters to the optimizer per _original_ parameter.

For what it is worth, we hope to default to `use_orig_params=True`, so please let us know if you hit any issues!

**Comment by rxqy** [comment]
@awgu I'try that! Many thanks for your help!

**Comment by ringohoffman** [comment]
@awgu is there any update on your work on this?

**Comment by mori360** [comment]
We are commited to support FSDP2. Here is the migration guide https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md
If you encounter any similar issues with FSDP2, please open a new issue with updated information.
